### ISSUE TITLE
Consider to integrate Postfixadmin to configure Postfix

### ISSUE THREAD

**Issue by neuffer** [ubermail/ubermail#5]
Postfixadmin should fit well with Ubermail I believe, it is a good admin interface and integrates with Postgres
https://github.com/postfixadmin/postfixadmin
http://postfixadmin.sourceforge.net/

**Comment by killua-eu** [comment]
Yep, I considered that already. This does have some issues though. Right now, the playbooks using the ubermail role are the "only source of truth" when it comes to domains, mailboxes/logins and passwords. This is a good thing for reproducibility (i.e. just run the role again during disaster recovery to get all accounts back just by playing the playbook) but bad for flexibility (users should be able to at least change their passwords). There are several ways how to deal with this:

- **full reproducibility, zero flexibility:** keep things as they are
- **full flexibility, zero reproducibility:** integrate postfixadmin, when running the playbook skip resetting user-changed passwords to initial ansible-defined defaults
- **something in between 1:** on each playbook run store the current mailboxes and password hashes in a hosts_file locally
- **something in between 2:** don't care about postfixadmin, just write a simple administrative app similar to postfixadmin that will generate the yml file and play the playbook
- **something in between 3:** figure out an account backup method that can be used together with the playbook's host_vars.

come from the top of my mind. Thoughts?

**Comment by neuffer** [comment]
On 10/23/18 8:34 PM, killua-eu wrote:
>
> Yep, I considered that already. This does have some issues though. 
> Right now, the playbooks using the ubermail role are the "only source 
> of truth" when it comes to domains, mailboxes/logins and passwords. 
> This is a good thing for reproducibility (i.e. just run the role again 
> during disaster recovery to get all accounts back just by playing the 
> playbook) but bad for flexibility (users should be able to at least 
> change their passwords). #
>
The "only source of truth" concept is a nightmare if you want to use the 
(mail)server for more then 2 domains and 2 users and need to 
update/upgrade it.

This is why sovereign fell through for me. An update/reinstall can not 
simply kill all my settings.
Sure, if I do a clean fresh install, defaults are fine. On an 
update/upgrade it is an absolute NO-GO
to simply wipe out all configurations, mail accounts, domains etc. That 
is a nightmare.

I have settings for about 50 domains, 2 dozen users, a handful of 
mailing lists and lots of email archives.

Losing the settings for those is not an option.

> There are several ways how to deal with this:
>
>   * *full reproducibility, zero flexibility:* keep things as they are
>   * *full flexibility, zero reproducibility:* integrate postfixadmin,
>     when running the playbook skip resetting user-changed passwords to
>     initial ansible-defined defaults
>
This does not only concern postfixadmin passwords. It also concerns 
(not) zapping all the tables in Postgres on update/upgrade and similar 
things.
>
>   * *something in between 1:* on each playbook run store the current
>     mailboxes and password hashes in a hosts_file locally
>   * *something in between 2:* don't care about postfixadmin, just
>     write a simple administrative app similar to postfixadmin that
>     will generate the yml file and play the playbook
>   * *something in between 3:* figure out an account backup method that
>     can be used together with the playbook's host_vars.
>
> come from the top of my mind. Thoughts?
>
A DNS server like PowerDNS would be nice to have as part as it supports 
Postgres as backend und supports DNS over TLS (|RFC7858|)/DNS over 
https,  etc.


<RANT>

Now imagine a new fix gets added to Ubermail github mainline and you run 
Ansible to apply the fix on your installation.

Suddenly all your mail accounts are gone (maybe even the mails if the 
Unix accounts get wiped), your DNS gets zapped, the VPN breaks down, 
family, friends, neighbors, customers can't reach you anymore or their 
mailboxes that they might have (had) on your server.

This is what sovereign/ubermail gives you at the moment with the "only 
source of truth" mindset.

That might be fine for a single person household when you really keep 
all your mail in Googlemail or something like that, but not if you need 
to keep the server up and running.

</RANT>


Postfixadmin has its own setup/upgrade script which doesn't wipe the DB 
tables, but setups/upgrades them from any previous version.

It checks a version number stored in the DB schema and applies updates 
as required to reach the current DDL version and without loosing the data.


For all other parts playbooks would need to be able to distinguish 
between update/upgrade and fresh install where necessary.

ie. configuring the time servers could be done from scratch, but 
anything that involves users, domains, mailboxes, certificates (linked 
to domains), etc. needs some kind of update/upgrade path. For Debian 
based systems this is already the default with proper packages anyway. 
When config files are updated the installer will check if the current 
file differs from the new one and ask for manual intervention if the 
install scripts do not handle the case.

This would give

  * *full reproducibility on install AND***flexibility on update/upgrade**

best of both worlds.

Comments on

  * something in between 1: That gets too complex too quickly as this is
    by far not only mailboxes and password hashes.
  * something in between 2: That is a LOT of work. The Postfixadmin guys
    have put into their interface what is needed to administrate
    postfix. Re-inventing the weel again and again doesn't make it better.
  * something in between 3: Except for the super admin account config I
    don't see anything there that should be stored on the host_vars, I'd
    pull DB exports that can be re-imported. Those as well as backups of
    the /etc/, /var/www, /var/mail, /home and maybe a few other
    directories would be the basis for a proper system backup.

What do you think?**
**

****

Cheers

   Mike

**Comment by killua-eu** [comment]
Hi Mike, thanks for your analysis.

- I don't really think sovereign zapped postgres tables on upgrades - I'll definitely doublecheck this later - but I'm quite sure that the role won't delete a thing. There are no tasks to remove unmanaged accounts or other settings, so if you encountered anything like this, something went wrong really badly for you.
- I definitely agree that having ansible as the only source of truth can be a pain if you wan't the smallest thing done just in a few clicks.
- Relying with upgrades on the packgemanager does't work, unless there'd be a whole "mailserver" package. There are dependencies between packages that evolve in time and that tend to break. Ansible is used for repetitive zero-knowledge deployments that just work, you can't really expect people to reconcile changes in configuration too.

As for my suggested solutions:

- something in between 1 is actually really easy. just the mailboxes, domains and hashes. Coupled with roles such as https://github.com/vaizard/mage-restic (for incremental, encrypted, deduplicated backups) you actually need just the above.
- something in between 2 is actually easy too, i've used postfixadmin for a long time myself, so I'm well aware how straightforward the app is. Its just that I dont know how many people run their playbooks from a deployment server, that would actually also host this ui.
- something in between 3 is actually hard(er), i think, since I cant think of an easy way to actually architect this in a predictable way.

**Comment by killua-eu** [comment]
To summarize, i'm not sure postfixadmin would be an easy-to-implement fit. But I'd quite inclined to resolve this issue somehow.

**Comment by killua-eu** [comment]
Actually you were right! Sovereign (and currently Ubermail as well) do both drop unmanaged users (in the templates/mailserver.sql.j2 file). Wow. I'll try to think of a proper fix, this is definitely a no-no.
